tempesta_fw/http.c

```diff
@@ -1070,23 +1126,48 @@ tfw_http_adjust_resp(TfwHttpResp *resp)
 	return 0;
 }
 
+/*
+ * Depending on results of processing of a request, either send the request
+ * to an appropriate server, or return the cached response. If none of that
+ * can be done for any reason, return HTTP 404 error to the client.
+ */
 static void
 tfw_http_req_cache_cb(TfwHttpReq *req, TfwHttpResp *resp, void *data)
 {
-	TfwConnection *conn = data;
+	int r;
 
 	if (resp) {
 		/*
 		 * We have prepared response, send it as is.
 		 * TODO should we adjust it somehow?
 		 */
-		tfw_connection_send(conn, (TfwMsg *)resp);
+		tfw_connection_send(req->conn, (TfwMsg *)resp);
 	} else {
-		if (tfw_http_adjust_req(req))
+		/* Dispatch request to an appropriate server. */
+		TfwConnection *conn = tfw_sched_get_srv_conn((TfwMsg *)req);
+		if (!conn)
+			goto send_404;
```

Comment:
This is not good and expected behavior (no cache and no upstream server), so please leave warning log message.